mqtt/mqtt-bridge/src/pump/builder.rs

```diff
@@ -0,0 +1,161 @@
+use std::convert::TryInto;
+
+use tokio::sync::mpsc;
+
+use crate::{
+    bridge::BridgeError,
+    client::{MqttClient, MqttClientConfig},
+    messages::{MessageHandler, TopicMapper},
+    persist::{PublicationStore, WakingMemoryStore},
+    settings::TopicRule,
+    upstream::{
+        LocalRpcHandler, LocalUpstreamHandler, LocalUpstreamPumpEventHandler, RemoteRpcHandler,
+        RemoteUpstreamHandler, RemoteUpstreamPumpEventHandler,
+    },
+};
+
+use super::{MessagesProcessor, Pump, PumpHandle};
+
+pub type PumpsResult = Result<
+    (
+        Pump<LocalUpstreamHandler<WakingMemoryStore>, LocalUpstreamPumpEventHandler>,
+        Pump<RemoteUpstreamHandler<WakingMemoryStore>, RemoteUpstreamPumpEventHandler>,
+    ),
+    BridgeError,
+>;
+
+pub struct Builder {
+    local: PumpBuilder,
+    remote: PumpBuilder,
+    store: Box<dyn Fn() -> PublicationStore<WakingMemoryStore>>,
+}
+
+impl Default for Builder {
+    fn default() -> Self {
+        Self {
+            local: PumpBuilder::default(),
+            remote: PumpBuilder::default(),
+            store: Box::new(|| PublicationStore::new_memory(0)),
+        }
+    }
+}
+
+impl Builder {
+    pub fn with_local<F>(&mut self, mut apply: F) -> &mut Self
```

Comment:
is there a case when the pump doesn't have both local and remote? I tough the with_ is used when there are cases when you can construct with/without

Reply:
no, there has to be both local and remote right now. there are some `.expect()` calls in the build method which will panic in runtime. not super rusty though...

Reply:
Then it means it should create them in the constructor instead?